clippy_lints/src/swap.rs

```diff
@@ -350,12 +350,24 @@ impl<'tcx> IndexBinding<'_, 'tcx> {
                 format!("{lhs_snippet}{rhs_snippet}")
             },
             ExprKind::Path(QPath::Resolved(_, path)) => {
-                let init = self.cx.expr_or_init(expr);
-
                 let Some(first_segment) = path.segments.first() else {
                     return String::new();
                 };
-                if !self.suggest_span.contains(init.span) || !self.is_used_other_than_swapping(first_segment.ident) {
+
+                let init = self.cx.expr_or_init(expr);
+
+                // We skip suggesting a variable binding in any of these cases:
+                // 1. Variable initialization is outside the suggestion span
+                // 2. Variable initialization is inside the suggestion span but the variable is not used as an index
+                //    or elsewhere later
+                // 3. Variable initialization is inside the suggestion span and the variable is used as an
+                //    index/elsewhere later, but its declaration is outside the suggestion span
+                if !self.suggest_span.contains(init.span)
+                    || !self.is_used_other_than_swapping(first_segment.ident)
```

Comment:
This test seems more costly than the new one (declaration outside the suggestion span), maybe you should move it in last position.